design/ceph/osd-migration.md

```diff
@@ -0,0 +1,61 @@
+
+
+## MOTIVATION:
+- Ceph uses BlueStore (since Luminous) as a special-purpose storage back end designed specifically for managing data on disk for Ceph OSD workloads. Ceph team is working on improving this backend storage that will help in significantly improving the performance, especially in disaster recovery scenarios. This new OSD backend is known as bluestore_rdr or seastore.
+- To support any new and improved storage backend, the existing OSD need to be destroyed, disks should be wiped out and new OSD should be created using the latest storage backend.
+
+## GOALS
+- Newly provisioned clusters should be able to use the specific storage backend while creating OSDs.
+- Existing clusters should be able to use a specific storage backend for OSDs with a minimal downtime and without losing their data.
+
+
+## PROPOSAL
+- Add a new field called `spec.Storage.osdStore` in the ceph cluster yaml.
+- This will contain:
+  - `spec.Storage.osdStore.type`: The type of backend store to be used for OSDs. For example,bluestore, bluestore_rdr, etc. The default store type will be `bluestore`
+  - `spec.Storage.osdStore.updateStore`: Whether the user wants to migrate the existing OSDs to use new storage backend. The migration process will involve destroying the exsisting OSDs one by one, wiping the disks, creating a new OSD and waiting for the PGs to be active+clean before migrating the next OSD. Since this operation involves possible downtime, users should be really sure before proceeding with this action. This field can only take the value as `yes-really-update-store`
+  -
+
+- Example:
+```yaml
+storage:
+  osdStore:
+    type: bluestore_rdr
+    updateStore: yes-really-update-store
+```
+- New deployments will use this `osdStore.type` while creating the OSDs.
+- If the user wants to change the `osdStore.type` field for an existing cluster, then they need to provide `spec.Storage.osdStore.updateStore:yes-really-update-store` in order to migrate the existing osds to use update osdStore.
+
+
+### NEW DEPLOYMENTS
+
+#### OSD Prepare
+- `ceph-volume prepare` command should be updated to use backend store provided in the `spec.storage.osdStore`. If no value is provided for `spec.storage.osdStore` then default to `bluestore` backend.
+
+  - **OSD on disks**:
+    - RAW MODE
+      - `ceph-volume lvm prepare --bluestore_rdr ...`
+    - LWM MODE
+      - `ceph-volume raw prepare --bluestore_rdr ...`
+  - **OSD on PVCs**:
+    -
+
+#### OSD Create
+- `OSD_STORE_FLAG` should use the value provided in the `spec.storage.osdStore`. If no values are provided by the user, then `OSD_STORE_FLAG` should default to `bluestore`
+- `ceph-volume` commands should use the `OSD_STORE_FLAG` while activating the OSDs.
+  - **OSD on disks**:
+    - RAW MODE
+    `ceph-volume raw activate "$OSD_STORE_FLAG" --device "$DEVICE" --no-systemd --no-tmpfs`
```

Comment:
The activate command requires the store type? I would hope the OSD activate could see on disk already what type it needs to work with. This seems risky if we pass the wrong store type to an existing OSD, instead of just passing the flag when provisioning a new OSD.

Reply:
yes. The activate command in [ceph-volume](https://github.com/rook/rook/blob/d20abd4aa739426ab23e9b4af530cf9fb6181da6/pkg/daemon/ceph/osd/daemon.go#L87-L88) is using the `storeType`

Reply:
Yes. We can skip OSD store flag in the activate command for both `lvm` and `raw` mode. More details [here](https://github.com/rook/rook/pull/12381#issuecomment-1668916773)